### ISSUE TITLE
Excessive memory consumption in mutations.

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#9873]
I think it was already fixed by #9860 (I can recheck after testing build will be published) but right now it reproduces well.
 
```
create table test_mutation_oom engine=MergeTree ORDER BY number as select number, randomPrintableASCII(1024) as logline, '' as string from numbers(10000000);

-- before mutation:
-- sudo pmap $(pidof -s clickhouse-server) | tail -n 1
--  total          3247180K

alter table test_mutation_oom update string = extract(logline, '(a[a-zA-Z]+)') where 1=1;

-- after mutation:
-- sudo pmap $(pidof -s clickhouse-server) | tail -n 1
-- total         27226188K
```

**Comment by filimonov** [comment]
Looks like this one is fixed earlier (by #9754)

**Comment by filimonov** [comment]
The case with complex PK is still broken (fixed in  #9860)
```
create table test_mutation_oom engine=MergeTree ORDER BY (number,n2) as select number, rand() as n2, randomPrintableASCII(1024) as logline, '' as string from numbers(10000000);
```

**Comment by filimonov** [comment]
Working properly now. (latest 20.3, 20.1, 20.4)
